File: Exponencial/Funcoes/quantil.m
function out = quantil(ordenado,original,q)

serie_intervalo=ordenado;
sinal_original=original;
serie_original=original;
n=length(serie_intervalo);
Q = q;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
t = zeros(n,1);
contador = 0;
for i = 1:n
    for j = 1:n
        if sinal_original(j)==serie_intervalo(i)
            t(i) = j;
            contador = contador +1;
            sinal_original(j) = power(10,12);
            break
        end
    end
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%determinando numero de quantis

q = zeros(Q-1,1);
for i = 1:Q-1 %atribuindo os valores dos quantis em um vetor
    q(i) = (n*(i))/Q+1;
    quantcomp = zeros(Q+1,1);
end
%q = q+1;
for i = 2:Q
    quantcomp(1) = 1;% quantcomp(0) = 0;
    quantcomp(i) = q(i-1);% quantcomp(i) = q(i-1);
    quantcomp(Q+1) = n+1;% quantcomp(Q) = length(serie_ordenada);
end

localiza = zeros(n,1);


for i = 1:n%adicionando cada elemento no seu quantil e atribuindo as posicoes dos quantis de cada elemento
    for j = 1:Q
        if i >= quantcomp(j,1) && i < quantcomp(j+1,1)
            localiza(t(i),1) = j;
            
        end
        if i == n
            localiza(t(i),1) = j;
        end
    end
end
%localiza(61,1)=20;
matrizadj = zeros(Q);
contador = 0;
for i = 1:(n-1)% construindo a matriz(ligando os quantis(nos))
 
    matrizadj(localiza(i,1),localiza(i+1,1)) = matrizadj(localiza(i,1),localiza(i+1,1)) + 1;
       contador = contador+1;
end

out =  matrizadj;
end

File: Python Figuras/Ciclos/quantil.py
# -*- coding: utf-8 -*-
"""
Created on Thu Jul 28 08:39:50 2022

@author: jgrel
"""

file_signal = open(r'/xlorenz.txt', 'r')
signal_original = []
for i in file_signal.readlines():
    line = i.strip('\n').split()
    signal_original.append(float(line[0]))
file_signal.close()

signal_sort = signal_original[:]

signal_sort.sort() #ordenando sinal
Q = 40 #determinando numero de quantis
   ##print("Numero de quantis : ", Q)
t = [ 0 for i in range(len(signal_sort))]

for i in range(len(signal_sort)):
    for j in range(len(signal_sort)):
        if(signal_original[j]== signal_sort[i]):
            t[i] = j
            signal_original[j]= pow(10,12)
            break
        
#         print("Vetor Posicao", t)
q = [0 for i in range(Q-1)]
for i in range((int(Q)-1)) : #atribuindo os valores dos quantis em um vetor
    q[i] = int((len(signal_sort)*(1+i))/int(Q))
    quantcomp = [ 0 for i in range(int(Q)+1)]
for i in range(1, int(Q)):
    quantcomp[0] =  0
    quantcomp[i]= q[i-1]
    quantcomp[Q]= int(len(signal_sort))
#         print("Vetor quantil", quantcomp)
localiza = [ 0 for i in range(len(signal_sort))]
contador = 0
for i in range(len(signal_sort)): #adicionando cada elemento no seu quantil e atribuindo as posicoes dos quantis de cada elemento
    for j in range(int(Q)):
        if (i >= quantcomp[j] and i < quantcomp[j+1]):
            localiza[t[i]] = j
            contador = contador +1
#         print(localiza)
matrizadj = []
matrizadj = [Q*[0] for i in range(int(Q))]
contador = 0
for i in range(len(signal_sort)-1): # ligando os quantis(nos)
    matrizadj[localiza[i]][localiza[i+1]] = matrizadj[localiza[i]][localiza[i+1]] + 1
    contador = contador +1
#         print("Matriz de adjacencia : ", matrizadj)
matriz_file = open(r'C:\Users\jgrel\Desktop\mestrado\CA\PY\matriz_adj.txt', 'w') #salvando arquivo txt

for i in range(len(matrizadj)):
    for j in range(len(matrizadj)):
        matriz_file.write('%d ' %matrizadj[i][j])
    matriz_file.write('\n')
matriz_file.close()